presto-orc/src/main/java/com/facebook/presto/orc/DictionaryCompressionOptimizer.java

```diff
@@ -106,30 +147,43 @@ public void reset()
     public void finalOptimize(int bufferedBytes)
     {
         updateDirectConversionCandidates();
-        convertLowCompressionStreams(bufferedBytes);
+        convertLowCompressionStreams(true, bufferedBytes);
     }
 
-    public void optimize(int bufferedBytes, int stripeRowCount)
+    @VisibleForTesting
+    boolean isUsefulCheckRequired(int dictionaryMemoryBytes)
     {
-        // recompute the dictionary memory usage
-        dictionaryMemoryBytes = allWriters.stream()
-                .filter(writer -> !writer.isDirectEncoded())
-                .mapToInt(DictionaryColumnManager::getDictionaryBytes)
-                .sum();
+        if (dictionaryMemoryBytes < dictionaryUsefulCheckColumnSizeBytes) {
+            return false;
+        }
 
-        // update the dictionary growth history
-        allWriters.stream()
-                .filter(writer -> !writer.isDirectEncoded())
-                .forEach(column -> column.updateHistory(stripeRowCount));
+        dictionaryUsefulCheckCounter++;
+        if (dictionaryUsefulCheckCounter == dictionaryUsefulCheckPerChunkFrequency) {
+            dictionaryUsefulCheckCounter = 0;
+            return true;
+        }
 
-        if (dictionaryMemoryBytes <= dictionaryMemoryMaxBytesLow) {
-            return;
+        return false;
+    }
+
+    public void optimize(int bufferedBytes, int stripeRowCount)
+    {
+        // recompute the dictionary memory usage
+        int totalDictionaryBytes = 0;
```

Comment:
It cannot go over Int.MAX, right?

Reply:
Previously we used integer to accumulate, so I preserve the behavior. There is possibility that it can go over INT_MAX, but that case would have many issues.